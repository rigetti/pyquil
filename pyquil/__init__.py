from pyquil._version import pyquil_version
from pyquil.quil import Program
from pyquil.api import list_quantum_computers, get_qc

__version__ = pyquil_version
