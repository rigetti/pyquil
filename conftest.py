import numpy as np
import pytest
from requests import RequestException

from pyquil.api import (
    QVMConnection,
    QVMCompiler,
    Client,
    BenchmarkConnection,
)
from pyquil.api._errors import UnknownApiError
from pyquil.api._abstract_compiler import QuilcNotRunning, QuilcVersionMismatch
from pyquil.api._qvm import QVMNotRunning, QVMVersionMismatch
from pyquil.device import Device
from pyquil.gates import I
from pyquil.paulis import sX
from pyquil.quil import Program
from pyquil.tests.utils import DummyCompiler


@pytest.fixture
def isa_dict():
    return {
        "1Q": {"0": {"type": "Xhalves"}, "1": {}, "2": {}, "3": {"dead": True}},
        "2Q": {
            "0-1": {},
            "1-2": {"type": "ISWAP"},
            "0-2": {"type": "CPHASE"},
            "0-3": {"dead": True},
        },
    }


@pytest.fixture
def specs_dict():
    return {
        "1Q": {
            "0": {
                "f1QRB": 0.99,
                "f1QRB_std_err": 0.01,
                "f1Q_simultaneous_RB": 0.98,
                "f1Q_simultaneous_RB_std_err": 0.02,
                "fRO": 0.93,
                "T1": 20e-6,
                "T2": 15e-6,
            },
            "1": {
                "f1QRB": 0.989,
                "f1QRB_std_err": 0.011,
                "f1Q_simultaneous_RB": 0.979,
                "f1Q_simultaneous_RB_std_err": 0.021,
                "fRO": 0.92,
                "T1": 19e-6,
                "T2": 12e-6,
            },
            "2": {
                "f1QRB": 0.983,
                "f1QRB_std_err": 0.017,
                "f1Q_simultaneous_RB": 0.973,
                "f1Q_simultaneous_RB_std_err": 0.027,
                "fRO": 0.95,
                "T1": 21e-6,
                "T2": 16e-6,
            },
            "3": {
                "f1QRB": 0.988,
                "f1QRB_std_err": 0.012,
                "f1Q_simultaneous_RB": 0.978,
                "f1Q_simultaneous_RB_std_err": 0.022,
                "fRO": 0.94,
                "T1": 18e-6,
                "T2": 11e-6,
            },
        },
        "2Q": {
            "0-1": {"fBellState": 0.90, "fCZ": 0.89, "fCZ_std_err": 0.01, "fCPHASE": 0.88},
            "1-2": {"fBellState": 0.91, "fCZ": 0.90, "fCZ_std_err": 0.12, "fCPHASE": 0.89},
            "0-2": {"fBellState": 0.92, "fCZ": 0.91, "fCZ_std_err": 0.20, "fCPHASE": 0.90},
            "0-3": {"fBellState": 0.89, "fCZ": 0.88, "fCZ_std_err": 0.03, "fCPHASE": 0.87},
        },
    }


@pytest.fixture
def noise_model_dict():
    return {
        "gates": [
            {
                "gate": "I",
                "params": (5.0,),
                "targets": (0, 1),
                "kraus_ops": [[[[1.0]], [[1.0]]]],
                "fidelity": 1.0,
            },
            {
                "gate": "RX",
                "params": (np.pi / 2.0,),
                "targets": (0,),
                "kraus_ops": [[[[1.0]], [[1.0]]]],
                "fidelity": 1.0,
            },
        ],
        "assignment_probs": {"1": [[1.0, 0.0], [0.0, 1.0]], "0": [[1.0, 0.0], [0.0, 1.0]]},
    }


@pytest.fixture
def device_raw(isa_dict, noise_model_dict, specs_dict):
    return {
        "isa": isa_dict,
        "noise_model": noise_model_dict,
        "specs": specs_dict,
        "is_online": True,
        "is_retuning": False,
    }


@pytest.fixture
def test_device(device_raw):
    return Device("test_device", device_raw)


@pytest.fixture(scope="session")
def qvm(client: Client):
    try:
        qvm = QVMConnection(client=client, random_seed=52)
        qvm.run(Program(I(0)), [])
        return qvm
    except (RequestException, QVMNotRunning, UnknownApiError) as e:
        return pytest.skip("This test requires QVM connection: {}".format(e))
    except QVMVersionMismatch as e:
        return pytest.skip("This test requires a different version of the QVM: {}".format(e))


@pytest.fixture()
def compiler(test_device, client: Client):
    try:
        compiler = QVMCompiler(device=test_device, client=client, timeout=1)
        compiler.quil_to_native_quil(Program(I(0)))
        return compiler
    except (RequestException, QuilcNotRunning, UnknownApiError, TimeoutError) as e:
        return pytest.skip("This test requires compiler connection: {}".format(e))
    except QuilcVersionMismatch as e:
        return pytest.skip("This test requires a different version of quilc: {}".format(e))


@pytest.fixture()
def dummy_compiler(test_device: Device, client: Client):
    return DummyCompiler(test_device, client)


@pytest.fixture(scope="session")
def client():
    return Client()


@pytest.fixture(scope="session")
def benchmarker(client: Client):
    try:
        bm = BenchmarkConnection(client=client, timeout=2)
        bm.apply_clifford_to_pauli(Program(I(0)), sX(0))
        return bm
    except (RequestException, TimeoutError) as e:
        return pytest.skip(
            "This test requires a running local benchmarker endpoint (ie quilc): {}".format(e)
        )


def _str_to_bool(s):
    """Convert either of the strings 'True' or 'False' to their Boolean equivalent"""
    if s == "True":
        return True
    elif s == "False":
        return False
    else:
        raise ValueError("Please specify either True or False")


def pytest_addoption(parser):
    parser.addoption(
        "--use-seed",
        action="store",
        type=_str_to_bool,
        default=True,
        help="run operator estimation tests faster by using a fixed random seed",
    )
    parser.addoption(
        "--runslow", action="store_true", default=False, help="run tests marked as being 'slow'"
    )


def pytest_configure(config):
    config.addinivalue_line("markers", "slow: mark test as slow to run")


def pytest_collection_modifyitems(config, items):
    if config.getoption("--runslow"):
        # --runslow given in cli: do not skip slow tests
        return
    skip_slow = pytest.mark.skip(reason="need --runslow option to run")
    for item in items:
        if "slow" in item.keywords:
            item.add_marker(skip_slow)


@pytest.fixture()
def use_seed(pytestconfig):
    return pytestconfig.getoption("use_seed")
