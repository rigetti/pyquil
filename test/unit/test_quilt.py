import os
from typing import Optional

import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from pyquil.gates import RX
from pyquil.quil import Program
from pyquil.quilatom import (
    FormalArgument,
    MemoryReference,
    Parameter,
)
from pyquil.quilbase import (
    AbstractInstruction,
    DefCalibration,
    DefMeasureCalibration,
    Gate,
    Qubit,
)
from pyquil.quiltcalibrations import (
    CalibrationMatch,
    match_calibration,
)
from pyquil.quiltwaveforms import (
    BoxcarAveragerKernel,
    DragGaussianWaveform,
    ErfSquareWaveform,
    FlatWaveform,
    GaussianWaveform,
    HrmGaussianWaveform,
)
from test.unit.conftest import TEST_DATA_DIR


def test_waveform_samples():
    # this is a very naive check: can we sample from the built-in template
    # waveforms?
    duration = 1e-6
    waveforms = [
        FlatWaveform(duration=duration, iq=1.0),
        FlatWaveform(duration=duration, iq=1.0 + 2.0j),
        GaussianWaveform(duration=duration, fwhm=2.0, t0=1.0),
        DragGaussianWaveform(duration=duration, fwhm=duration / 4, t0=duration / 2, anh=5.0, alpha=3.0),
        HrmGaussianWaveform(
            duration=duration,
            fwhm=duration / 4,
            t0=duration / 2,
            anh=5.0,
            alpha=3.0,
            second_order_hrm_coeff=0.5,
        ),
        ErfSquareWaveform(duration=duration, risetime=duration / 8, pad_left=0.0, pad_right=0.0),
        BoxcarAveragerKernel(duration=duration),
    ]

    rates = [int(1e9), 1e9, 1e9 + 0.5]

    for rate in rates:
        for wf in waveforms:
            assert wf.samples(rate) is not None


def test_waveform_samples_optional_args():
    def flat(**kwargs):
        return FlatWaveform(duration=1e-8, iq=1.0, **kwargs).samples(1e9)

    assert np.array_equal(2.0 * flat(), flat(scale=2.0))
    assert np.array_equal(np.exp(1j) * flat(), flat(phase=1.0))


def _match(cal: DefCalibration, instr: AbstractInstruction) -> Optional[CalibrationMatch]:
    # get the first line, remove trailing colon if present
    cal_header = cal.splitlines()[0].strip().replace(":", "")
    # convert to quil ast
    full_calibration = cal_header + ":\n    NOP\n\n"
    cal = Program(full_calibration).calibrations[0]

    # we pull the last instr (omitting implicit DECLARE)
    instr = Program(instr)[-1]
    return match_calibration(instr, cal)


def _match_measure(cal: DefMeasureCalibration, instr: AbstractInstruction) -> Optional[CalibrationMatch]:
    # get the first line, remove trailing colon if present
    cal_header = cal.splitlines()[0].strip().replace(":", "")
    # convert to quil ast
    full_calibration = cal_header + ":\n    NOP\n\n"
    cal = Program(full_calibration).measure_calibrations[0]

    # we pull the last instr (omitting implicit DECLARE)
    instr = Program(instr)[-1]
    return match_calibration(instr, cal)


def test_simple_gate_calibration_match():
    matches = [
        ("DEFCAL X 0", "X 0"),
        ("DEFCAL X q", "X 0"),
        ("DEFCAL FOO q 0", "FOO 1 0"),
        ("DEFCAL FOO q 0", "FOO 0 0"),
    ]

    for cal, instr in matches:
        assert _match(cal, instr) is not None

    assert _match(cal, instr).settings[FormalArgument("q")] == Qubit(0)

    mismatches = [
        ("DEFCAL X 0", "X 1"),
        ("DEFCAL Y 0", "X 0"),
        # we are case sensitive
        ("DEFCAL foo 0", "FOO 0"),
    ]

    for cal, instr in mismatches:
        assert _match(cal, instr) is None


def test_parametric_calibration_match():
    matches = [
        ("DEFCAL RX(0.0) 0", "RX(0.0) 0"),
        ("DEFCAL RX(0.0) 0", "RX(0*pi) 0"),
        ("DEFCAL RX(pi/2) 0", "RX(pi/2) 0"),
        ("DEFCAL RX(pi/2) q", "RX(pi/2) 0"),
        ("DEFCAL RX(pi/2) q", "RX(pi/2) 1"),
        ("DEFCAL RZ(pi/2) 0", "RZ(pi/2) 0"),
        ("DEFCAL RX(%theta) 0", "RX(pi) 0"),
    ]

    for cal, instr in matches:
        assert _match(cal, instr) is not None

    assert np.isclose(_match(cal, instr).settings[Parameter("theta")], np.pi)

    mismatches = [
        ("DEFCAL RX(pi) q", "RX(0) 0"),
        ("DEFCAL RX(pi) 0", "RX(0) 0"),
        ("DEFCAL RX(pi) 0", "RY(pi) 0"),
    ]

    for cal, instr in mismatches:
        assert not _match(cal, instr)


def test_memory_reference_parameter():
    assert _match("DEFCAL RX(%theta) q", "DECLARE theta REAL\nRZ(theta) 0") is None


def test_measure_calibration_match():
    matches = [
        ("DEFCAL MEASURE 0 dest", "MEASURE 0 ro[0]"),
        ("DEFCAL MEASURE q dest", "MEASURE 0 ro[1]"),
        ("DEFCAL MEASURE 0 b", "DECLARE ro BIT\nMEASURE 0 ro[0]"),
        ("DEFCAL MEASURE q b", "DECLARE ro BIT\nMEASURE 1 ro[0]"),
    ]

    for cal, instr in matches:
        assert _match_measure(cal, instr) is not None

    match = _match_measure(cal, instr)
    assert match.settings[FormalArgument("q")] == Qubit(1)
    assert match.settings[MemoryReference("b")] == MemoryReference("ro")

    mismatches = [
        ("DEFCAL MEASURE 1 dest", "MEASURE 0"),
        ("DEFCAL MEASURE 0 b", "MEASURE 0"),
    ]

    for cal, instr in mismatches:
        assert _match_measure(cal, instr) is None


def test_program_match_last():
    first = DefCalibration("X", [], [Qubit(0)], [RX(np.pi / 2, 0)])
    second = DefCalibration("X", [], [Qubit(0)], [RX(-np.pi / 2, 0)])
    prog = Program(first, second)
    match = prog.match_calibrations(Gate("X", [], [Qubit(0)]))
    assert match.cal.out() == second.out()
    assert match.settings == {}


@pytest.mark.parametrize(
    "program_input,gate,program_output",
    [
        (
            Program(
                """
DEFCAL RZ(%theta) q:
    SHIFT-PHASE q "rf" -%theta
"""
            ),
            Gate("RZ", [np.pi], [Qubit(0)]),
            Program('SHIFT-PHASE 0 "rf" -3.141592653589793'),
        ),
        (
            Program(
                """
DEFCAL A(%theta) q:
    SHIFT-PHASE q "rf" -%theta

DEFCAL RZ(%theta) q:
    SHIFT-PHASE q "rf" -%theta
    A(%theta) q
"""
            ),
            Gate("RZ", [np.pi], [Qubit(0)]),
            Program('SHIFT-PHASE 0 "rf" -3.141592653589793', 'SHIFT-PHASE 0 "rf" -3.141592653589793'),
        ),
        (
            Program(
                """
DEFCAL A(%theta) q:
    RX(%theta) q

DEFCAL B(%theta) q:
    RY(%theta) q
DEFCAL RZ(%theta) q:
    A(%theta) q
    B(%theta) q
    A(%theta) q
"""
            ),
            Gate("RZ", [np.pi], [Qubit(0)]),
            Program("RX(3.141592653589793) 0", "RY(3.141592653589793) 0", "RX(3.141592653589793) 0"),
        ),
        (
            Program(
                """
DEFCAL RX(%theta) q:
    RY(%theta) q

DEFCAL RZ(%theta) q:
    RX(%theta) q
"""
            ),
            Gate("RZ", [np.pi], [Qubit(0)]),
            Program("RY(3.141592653589793) 0"),
        ),
        (
            Program(
                """
DEFCAL RY(pi) 0:
    I 0
DEFCAL RZ(0) 0:
    RY(pi) 0
DEFCAL RX(0) 0:
    RY(pi) 0
    RZ(0) 0
"""
            ),
            Gate("RX", [0], [Qubit(0)]),
            Program("I 0", "I 0"),
        ),
        (
            Program(
                """
DEFCAL RX(%theta) 0:
    RX(%theta / 2) 0

DEFCAL RX(pi / 4) 0:
    I 0
"""
            ),
            Gate("RX", [np.pi], [Qubit(0)]),
            Program("I 0"),
        ),
    ],
)
def test_program_calibrate(program_input, gate, program_output):
    calibrated = program_input.calibrate(gate)
    assert Program(calibrated).out() == program_output.out()


def test_program_calibrate_no_match():
    program = Program()
    instruction = Gate("RX", [np.pi], [Qubit(0)])
    assert program.calibrate(instruction) == [instruction]


@pytest.mark.parametrize(
    "program_text",
    (
        """
DEFCAL RZ(%theta) q:
    RZ(%theta) q
""",
        """
DEFCAL RX(%theta) q:
    RZ(%theta) q

DEFCAL RZ(%theta) q:
    RX(%theta) q
""",
        """
DEFCAL RX(%theta) q:
    RZ(0) q

DEFCAL RZ(%theta) q:
    RX(%theta) q
""",
        """
DEFCAL RX(%theta) q:
    RZ(%theta) q

DEFCAL RZ(%theta) q:
    RX(0) q
""",
        """
DEFCAL RZ(%theta) q:
    RZ(0) q

DEFCAL RZ(0) q:
    RZ(pi) q
""",
    ),
)
def test_program_calibrate_cyclic_error(program_text):
    prog = Program(program_text)
    with pytest.raises(Exception):
        prog.calibrate(Gate("RZ", [np.pi], [Qubit(0)]))


def test_merge_programs_with_quilt_features():
    prog_1 = Program(
        """
DEFCAL RZ(%theta) q:
    SHIFT-PHASE q "rf" -%theta
"""
    )
    prog_2 = Program() + prog_1
    assert len(prog_2.calibrations) == 1
    prog_2 = prog_1 + Program()
    assert len(prog_2.calibrations) == 1
    prog_2 = Program()
    prog_2 += prog_1
    assert len(prog_2.calibrations) == 1

    prog_1 = Program(
        """
DEFFRAME 0 "rf":
    SAMPLE-RATE: 2.0
"""
    )
    prog_2 = Program() + prog_1
    assert len(prog_2.frames) == 1
    prog_2 = prog_1 + Program()
    assert len(prog_2.frames) == 1
    prog_2 = Program()
    prog_2 += prog_1
    assert len(prog_2.frames) == 1

    prog_1 = Program(
        """
DEFWAVEFORM foo:
    1.0, 1.0, 1.0
"""
    )
    prog_2 = Program() + prog_1
    assert len(prog_2.waveforms) == 1
    prog_2 = prog_1 + Program()
    assert len(prog_2.waveforms) == 1
    prog_2 = Program()
    prog_2 += prog_1
    assert len(prog_2.waveforms) == 1


@pytest.mark.parametrize("calibration_program_suffix", ["rx", "cz", "cz_cphase", "xy", "measure"])
def test_round_trip_calibration_program(calibration_program_suffix: str, snapshot: SnapshotAssertion):
    """Test we round-trip a multi-part Quil-T calibration program; tests calibration overlap elimination."""
    with open(os.path.join(TEST_DATA_DIR, f"calibration_program_{calibration_program_suffix}.quil")) as file:
        calibration_program_text = file.read()

    calibration_program = Program(calibration_program_text)
    assert calibration_program.out() == snapshot
