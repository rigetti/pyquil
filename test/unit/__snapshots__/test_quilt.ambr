# serializer version: 1
# name: test_round_trip_calibration_program[cz]
  '''
  DEFCAL CZ 110 111:
  	FENCE
  	NONBLOCKING PULSE 110 111 "cz" q110_q111_cz/CZ
  	DELAY 110 "rf" 1.6e-7
  	SHIFT-PHASE 110 "rf" 0.6
  	DELAY 111 "rf" 1.6e-7
  	SHIFT-PHASE 111 "rf" 0.4
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.3
  	SHIFT-PHASE 110 117 "xy" 0.3
  	SHIFT-PHASE 110 111 "xy" -0.2
  	SHIFT-PHASE 111 112 "xy" -0.2
  	SHIFT-PHASE 111 126 "xy" -0.2
  DEFCAL CZ 111 110:
  	FENCE
  	NONBLOCKING PULSE 110 111 "cz" q110_q111_cz/CZ
  	DELAY 110 "rf" 1.6e-7
  	SHIFT-PHASE 110 "rf" 0.6
  	DELAY 111 "rf" 1.6e-7
  	SHIFT-PHASE 111 "rf" 0.4
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.3
  	SHIFT-PHASE 110 117 "xy" 0.3
  	SHIFT-PHASE 110 111 "xy" -0.2
  	SHIFT-PHASE 111 112 "xy" -0.2
  	SHIFT-PHASE 111 126 "xy" -0.2
  
  '''
# ---
# name: test_round_trip_calibration_program[cz_cphase]
  '''
  DEFCAL CZ 110 111:
  	FENCE
  	NONBLOCKING PULSE 110 111 "cz" q110_q111_cz/CZ
  	DELAY 110 "rf" 1.6e-7
  	SHIFT-PHASE 110 "rf" 0.6
  	DELAY 111 "rf" 1.6e-7
  	SHIFT-PHASE 111 "rf" 0.4
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.3
  	SHIFT-PHASE 110 117 "xy" 0.3
  	SHIFT-PHASE 110 111 "xy" -0.2
  	SHIFT-PHASE 111 112 "xy" -0.2
  	SHIFT-PHASE 111 126 "xy" -0.2
  DEFCAL CZ 111 110:
  	FENCE
  	NONBLOCKING PULSE 110 111 "cz" q110_q111_cz/CZ
  	DELAY 110 "rf" 1.6e-7
  	SHIFT-PHASE 110 "rf" 0.6
  	DELAY 111 "rf" 1.6e-7
  	SHIFT-PHASE 111 "rf" 0.4
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.3
  	SHIFT-PHASE 110 117 "xy" 0.3
  	SHIFT-PHASE 110 111 "xy" -0.2
  	SHIFT-PHASE 111 112 "xy" -0.2
  	SHIFT-PHASE 111 126 "xy" -0.2
  DEFCAL CPHASE(%theta) 110 111:
  	FENCE
  	NONBLOCKING PULSE 110 111 "cphase" q110_q111_cphase/sqrtCPHASE
  	SHIFT-PHASE 110 111 "cphase" -1.2+(0.5*%theta)
  	NONBLOCKING PULSE 110 111 "cphase" q110_q111_cphase/sqrtCPHASE
  	SHIFT-PHASE 110 111 "cphase" -1*(-1.2+(0.5*%theta))
  	DELAY 110 "rf" 3.2e-7
  	SHIFT-PHASE 110 "rf" 2.2
  	DELAY 111 "rf" 3.2e-7
  	SHIFT-PHASE 111 "rf" 1.2
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 1.1
  	SHIFT-PHASE 110 117 "xy" 1.1
  	SHIFT-PHASE 110 111 "xy" -0.6
  	SHIFT-PHASE 111 112 "xy" -0.6
  	SHIFT-PHASE 111 126 "xy" -0.6
  DEFCAL CPHASE(%theta) 111 110:
  	FENCE
  	NONBLOCKING PULSE 110 111 "cphase" q110_q111_cphase/sqrtCPHASE
  	SHIFT-PHASE 110 111 "cphase" -1.2+(0.5*%theta)
  	NONBLOCKING PULSE 110 111 "cphase" q110_q111_cphase/sqrtCPHASE
  	SHIFT-PHASE 110 111 "cphase" -1*(-1.2+(0.5*%theta))
  	DELAY 110 "rf" 3.2e-7
  	SHIFT-PHASE 110 "rf" 2.2
  	DELAY 111 "rf" 3.2e-7
  	SHIFT-PHASE 111 "rf" 1.2
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 1.1
  	SHIFT-PHASE 110 117 "xy" 1.1
  	SHIFT-PHASE 110 111 "xy" -0.6
  	SHIFT-PHASE 111 112 "xy" -0.6
  	SHIFT-PHASE 111 126 "xy" -0.6
  
  '''
# ---
# name: test_round_trip_calibration_program[measure]
  '''
  DEFCAL MEASURE 0 addr:
  	FENCE 0
  	DECLARE q0_unclassified REAL[2]
  	PULSE 0 "rf_f12" gaussian(detuning: 0, duration: 6e-8, fwhm: 1.6e-8, phase: 0, scale: 0.5, t0: 3e-8)
  	NONBLOCKING PULSE 0 "ro_tx" flat(detuning: 0, duration: 1.6e-6, iq: 1, phase: 0, scale: 0.1)
  	NONBLOCKING CAPTURE 0 "ro_rx" boxcar_kernel(detuning: 0, duration: 1.6e-6, phase: 0, scale: 1) q0_unclassified[0]
  	PRAGMA FILTER-NODE q0_unclassified "{'module':'standard.filters.io','filter_type':'DataBuffer','source':'q0_ro_rx/filter','publish':true,'params':{},'_type':'FilterNode'}"
  	PRAGMA LOAD-MEMORY q0_unclassified "q0_unclassified[0]"
  	PRAGMA FILTER-NODE q0_classified "{'module':'standard.filters.classifiers','filter_type':'SingleQLinear','source':'q0_ro_rx/filter','publish':false,'params':{'a':[1.0,0.0],'threshold':0.01},'_type':'FilterNode'}"
  	PRAGMA FILTER-NODE q0 "{'module':'standard.filters.io','filter_type':'DataBuffer','source':'q0_classified','publish':true,'params':{},'_type':'FilterNode'}"
  	PRAGMA LOAD-MEMORY q0 "addr"
  	FENCE 0
  
  DEFCAL MEASURE 10 addr:
  	FENCE 10
  	DECLARE q10_unclassified REAL[2]
  	PULSE 10 "rf_f12" gaussian(detuning: 0, duration: 7e-8, fwhm: 1.7e-8, phase: 0, scale: 0.5, t0: 3e-8)
  	NONBLOCKING PULSE 10 "ro_tx" flat(detuning: 0, duration: 1.6e-6, iq: 1, phase: 0, scale: 0.1)
  	NONBLOCKING CAPTURE 10 "ro_rx" boxcar_kernel(detuning: 0, duration: 1.6e-6, phase: 0, scale: 1) q10_unclassified[0]
  	PRAGMA FILTER-NODE q10_unclassified "{'module':'standard.filters.io','filter_type':'DataBuffer','source':'q10_ro_rx/filter','publish':true,'params':{},'_type':'FilterNode'}"
  	PRAGMA LOAD-MEMORY q10_unclassified "q10_unclassified[0]"
  	PRAGMA FILTER-NODE q10_classified "{'module':'standard.filters.classifiers','filter_type':'SingleQLinear','source':'q10_ro_rx/filter','publish':false,'params':{'a':[1.0,0.0],'threshold':0.01},'_type':'FilterNode'}"
  	PRAGMA FILTER-NODE q10 "{'module':'standard.filters.io','filter_type':'DataBuffer','source':'q10_classified','publish':true,'params':{},'_type':'FilterNode'}"
  	PRAGMA LOAD-MEMORY q10 "addr"
  	FENCE 10
  
  DEFCAL MEASURE 100 addr:
  	FENCE 100
  	DECLARE q100_unclassified REAL[2]
  	PULSE 100 "rf_f12" gaussian(detuning: 0, duration: 8e-8, fwhm: 1.8e-8, phase: 0, scale: 0.5, t0: 3e-8)
  	NONBLOCKING PULSE 100 "ro_tx" flat(detuning: 0, duration: 1.6e-6, iq: 1, phase: 0, scale: 0.1)
  	NONBLOCKING CAPTURE 100 "ro_rx" boxcar_kernel(detuning: 0, duration: 1.6e-6, phase: 0, scale: 1) q100_unclassified[0]
  	PRAGMA FILTER-NODE q100_unclassified "{'module':'standard.filters.io','filter_type':'DataBuffer','source':'q100_ro_rx/filter','publish':true,'params':{},'_type':'FilterNode'}"
  	PRAGMA LOAD-MEMORY q100_unclassified "q100_unclassified[0]"
  	PRAGMA FILTER-NODE q100_classified "{'module':'standard.filters.classifiers','filter_type':'SingleQLinear','source':'q10_ro_rx/filter','publish':false,'params':{'a':[1.0,0.0],'threshold':0.01},'_type':'FilterNode'}"
  	PRAGMA FILTER-NODE q100 "{'module':'standard.filters.io','filter_type':'DataBuffer','source':'q100_classified','publish':true,'params':{},'_type':'FilterNode'}"
  	PRAGMA LOAD-MEMORY q100 "addr"
  	FENCE 100
  
  
  '''
# ---
# name: test_round_trip_calibration_program[rx]
  '''
  DEFCAL RX(pi/2) 0:
  	FENCE 0
  	NONBLOCKING PULSE 0 "rf" drag_gaussian(alpha: -0.1, anh: -100000000, detuning: 0, duration: 1e-8, fwhm: 1e-8, phase: 0, scale: 0.1, t0: 1e-8)
  	FENCE 0
  DEFCAL RX(-pi/2) 0:
  	FENCE 0
  	NONBLOCKING PULSE 0 "rf" drag_gaussian(alpha: -0.2, anh: -100000000, detuning: 0, duration: 1e-8, fwhm: 1e-8, phase: 0, scale: 0.2, t0: 1e-8)
  	FENCE 0
  DEFCAL RX(pi) 0:
  	FENCE 0
  	NONBLOCKING PULSE 0 "rf" drag_gaussian(alpha: -0.3, anh: -100000000, detuning: 0, duration: 1e-8, fwhm: 1e-8, phase: 0, scale: 0.3, t0: 1e-8)
  	FENCE 0
  DEFCAL RX(-pi) 0:
  	FENCE 0
  	NONBLOCKING PULSE 0 "rf" drag_gaussian(alpha: -0.4, anh: -100000000, detuning: 0, duration: 1e-8, fwhm: 1e-8, phase: 0, scale: 0.4, t0: 1e-8)
  	FENCE 0
  DEFCAL RX(pi/2) 3:
  	FENCE 3
  	NONBLOCKING PULSE 3 "rf" drag_gaussian(alpha: -0.1, anh: -100000000, detuning: 0, duration: 2e-8, fwhm: 1e-8, phase: 0, scale: 0.1, t0: 1e-8)
  	FENCE 3
  DEFCAL RX(-pi/2) 3:
  	FENCE 3
  	NONBLOCKING PULSE 3 "rf" drag_gaussian(alpha: -0.2, anh: -100000000, detuning: 0, duration: 2e-8, fwhm: 1e-8, phase: 0, scale: 0.2, t0: 1e-8)
  	FENCE 3
  DEFCAL RX(pi) 3:
  	FENCE 3
  	NONBLOCKING PULSE 3 "rf" drag_gaussian(alpha: -0.3, anh: -100000000, detuning: 0, duration: 2e-8, fwhm: 1e-8, phase: 0, scale: 0.3, t0: 1e-8)
  	FENCE 3
  DEFCAL RX(-pi) 3:
  	FENCE 3
  	NONBLOCKING PULSE 3 "rf" drag_gaussian(alpha: -0.4, anh: -100000000, detuning: 0, duration: 2e-8, fwhm: 1e-8, phase: 0, scale: 0.4, t0: 1e-8)
  	FENCE 3
  
  '''
# ---
# name: test_round_trip_calibration_program[xy]
  '''
  DEFCAL XY(%theta) 110 111:
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.4
  	NONBLOCKING PULSE 110 111 "xy" q110_q111_xy/sqrtiSWAP
  	SHIFT-PHASE 110 111 "xy" -0.4
  	SHIFT-PHASE 110 111 "xy" 1.8+(-0.5*%theta)
  	NONBLOCKING PULSE 110 111 "xy" q110_q111_xy/sqrtiSWAP
  	SHIFT-PHASE 110 111 "xy" -1*(1.8+(-0.5*%theta))
  	DELAY 110 "rf" 1.2e-7
  	SHIFT-PHASE 110 "rf" 0.8
  	DELAY 111 "rf" 1.2e-7
  	SHIFT-PHASE 111 "rf" 0.4
  	SHIFT-PHASE 111 "rf" 0.5*%theta
  	SHIFT-PHASE 110 "rf" -0.5*%theta
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.4
  	SHIFT-PHASE 110 117 "xy" 0.4
  	SHIFT-PHASE 110 111 "xy" -0.2
  	SHIFT-PHASE 111 112 "xy" -0.2
  	SHIFT-PHASE 111 126 "xy" -0.2
  	SHIFT-PHASE 110 111 "xy" -0.25*%theta
  	SHIFT-PHASE 111 112 "xy" -0.25*%theta
  	SHIFT-PHASE 111 126 "xy" -0.25*%theta
  	SHIFT-PHASE 110 111 "xy" -0.25*%theta
  	SHIFT-PHASE 110 117 "xy" -0.25*%theta
  DEFCAL XY(%theta) 111 110:
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.4
  	NONBLOCKING PULSE 110 111 "xy" q110_q111_xy/sqrtiSWAP
  	SHIFT-PHASE 110 111 "xy" -0.4
  	SHIFT-PHASE 110 111 "xy" 1.8+(-0.5*%theta)
  	NONBLOCKING PULSE 110 111 "xy" q110_q111_xy/sqrtiSWAP
  	SHIFT-PHASE 110 111 "xy" -1*(1.8+(-0.5*%theta))
  	DELAY 110 "rf" 1.2e-7
  	SHIFT-PHASE 110 "rf" 0.8
  	DELAY 111 "rf" 1.2e-7
  	SHIFT-PHASE 111 "rf" 0.4
  	SHIFT-PHASE 111 "rf" 0.5*%theta
  	SHIFT-PHASE 110 "rf" -0.5*%theta
  	FENCE
  	SHIFT-PHASE 110 111 "xy" 0.4
  	SHIFT-PHASE 110 117 "xy" 0.4
  	SHIFT-PHASE 110 111 "xy" -0.2
  	SHIFT-PHASE 111 112 "xy" -0.2
  	SHIFT-PHASE 111 126 "xy" -0.2
  	SHIFT-PHASE 110 111 "xy" -0.25*%theta
  	SHIFT-PHASE 111 112 "xy" -0.25*%theta
  	SHIFT-PHASE 111 126 "xy" -0.25*%theta
  	SHIFT-PHASE 110 111 "xy" -0.25*%theta
  	SHIFT-PHASE 110 117 "xy" -0.25*%theta
  
  '''
# ---
